MobileBuy/sample/src/main/java/com/shopify/sample/view/products/ProductListView.java

```diff
@@ -40,79 +37,88 @@
 
-public final class ProductListView extends SwipeRefreshLayout implements LifecycleOwner, RecyclerViewAdapter.OnItemClickListener {
   @BindView(R.id.list) RecyclerView listView;
 
-  private final LifecycleRegistry lifecycleRegistry = new LifecycleRegistry(this);
-  private final RecyclerViewAdapter listViewAdapter = new RecyclerViewAdapter(this);
-  private BasePaginatedListViewModel<Product> viewModel;
+  private RecyclerViewAdapter adapter;
+  private ProductListViewModel viewModel;
 
-  public ProductListView(@NonNull final Context context) {
+  public ProductListView(final Context context) {
     super(context);
-    lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE);
   }
 
-  public ProductListView(@NonNull final Context context, @Nullable final AttributeSet attrs) {
+  public ProductListView(final Context context, final AttributeSet attrs) {
     super(context, attrs);
-    lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE);
   }
 
-  @Override public Lifecycle getLifecycle() {
-    return lifecycleRegistry;
+  @Override
+  public void onItemClick(@NonNull final ListItemViewModel itemViewModel) {
+    if (itemViewModel.payload() instanceof Product) {
+      ScreenRouter.route(getContext(), new ProductClickActionEvent((Product) itemViewModel.payload()));
+    }
   }
 
-  public void bindViewModel(@NonNull final BasePaginatedListViewModel<Product> viewModel) {
-    this.viewModel = checkNotNull(viewModel, "viewModel == null");
+  @Override
+  public void onNextPage() {
+    viewModel.fetchData();
+  }
 
+  @Override
+  public void onRefresh() {
     viewModel.reset();
-    viewModel.progressLiveData().observe(this, progress -> {
-      if (progress != null) {
-        setRefreshing(progress.show);
-      }
-    });
-    viewModel.errorErrorCallback().observe(this, error -> {
-      if (error != null) {
-        showDefaultErrorMessage();
-      }
-    });
-    viewModel.listItemsLiveData().observe(this, this::swapItems);
+    viewModel.fetchData();
   }
 
-  @Override protected void onFinishInflate() {
+  public void bindViewModel(@NonNull final ProductListViewModel viewModel) {
+    this.viewModel = viewModel;
+    viewModel.fetchDataIfNecessary();
+    viewModel
+      .state()
+      .observe(this, state -> setRefreshing(state == CollectionListViewModel.State.FETCHING));
+    viewModel
+      .error()
+      .observe(this, error -> {
+        Snackbar snackbar = Snackbar.make(this, R.string.default_error, Snackbar.LENGTH_LONG);
+        snackbar.getView().setBackgroundResource(R.color.snackbar_error_background);
+        snackbar.show();
+      });
+    Transformations
+      .map(viewModel.data(), products -> Util.reduce(products, (viewModels, product) -> {
+        viewModels.add(new ProductListItemViewModel(product));
```

Comment:
Is there any reason why this transformation can be done on ViewModel side?
So to have  `public class ProductListViewModel extends BasePaginatedListViewModel< ProductListItemViewModel >`